Source/WTF/wtf/ThreadSafeWeakHashSet.h

```diff
@@ -171,27 +164,36 @@ class ThreadSafeWeakHashSet final {
             callback(item.get());
     }
 
+    unsigned sizeIncludingEmptyEntriesForTesting()
+    {
+        Locker locker { m_lock };
+        return m_map.size();
+    }
+
 private:
     ALWAYS_INLINE void moveFrom(ThreadSafeWeakHashSet&& other)
     {
         Locker locker { m_lock };
         Locker otherLocker { other.m_lock };
-        m_set = std::exchange(other.m_set, { });
+        m_map = std::exchange(other.m_map, { });
         m_operationCountSinceLastCleanup = std::exchange(other.m_operationCountSinceLastCleanup, 0);
     }
 
+    static constexpr unsigned initialMaxOperationCountWithoutCleanup = 512;
     ALWAYS_INLINE void amortizedCleanupIfNeeded() const WTF_REQUIRES_LOCK(m_lock)
     {
-        if (++m_operationCountSinceLastCleanup / 2 > m_set.size()) {
-            m_set.removeIf([] (auto& pair) {
-                return pair.first->objectHasStartedDeletion();
+        if (++m_operationCountSinceLastCleanup / 2 > m_map.size() || m_operationCountSinceLastCleanup > m_maxOperationCountWithoutCleanup) {
+            bool didRemove = m_map.removeIf([] (auto& pair) {
+                return pair.value->objectHasStartedDeletion();
             });
+            m_maxOperationCountWithoutCleanup = didRemove ? initialMaxOperationCountWithoutCleanup : m_maxOperationCountWithoutCleanup * 2;
```

Comment:
Though unlikely to be hit in practice, I think jumping down to initialMaxOperationCountWithoutCleanup if there's a removal can lead to O(n^2) algorithms when removing lots of objects from the set.  I think it would be better to jump down to std::max(initialMaxOperationCountWithoutCleanup, m_maxOperationCountWithoutCleanup / 2) would be better.